v0.29: add new search parameter `matchingStrategy` (#1865)
meilisearch/documentation: reference/api/search.md

```

[You can read more about location-based sorting in our dedicated guide.](/learn/advanced/geosearch.md#sorting-results-with-geopoint)

### Matching strategy

**Parameter**: `matchingStrategy`
**Expected value**: `last` or `all`
**Default value**: `last`

Defines the strategy used to match query terms in documents. `last` returns documents containing all the query words first. If there aren't enough documents to meet the requested `limit`, it will remove one query term at a time, starting from the end of the query.

<CodeSamples id="search_parameter_guide_matching_strategy_1" />

With the above code sample, Meilisearch will first return documents that contain all three words. If the results don't meet the requested `limit`, it will return documents with just `big fat` followed by just `big`.

> `all`, on the other hand, only returns documents that contain all query terms. Meilisearch will not match any more documents even if there aren't enough to meet the requested `limit`. Searching for `big fat liar` will only return documents that contain all three words.

Reviewer: We want to avoid ambiguous constructions like "on the other hand" in our API reference. We don't know what the reader may have read before this: he or she could be speed-reading, or could have followed a link directly to this paragraph. Either way they would be missing the necessary context to understand what we mean by "on the other hand". (Not to mention that we may not always have only two values for this parameter)

```suggestion
#### `all`

`all` only returns documents that contain all query terms. Meilisearch will not match any more documents even if there aren't enough to meet the requested `limit`. 
```

Reviewer 2: Thanks for the tips Tommy!

<CodeSamples id="search_parameter_guide_matching_strategy_2" />

> `all` usually returns fewer search results compared to `last`.

Reviewer: Makes more sense to give this sentence after the code sample.

```suggestion
The above code sample would only return documents containing all three words.
```

```

[You can read more about location-based sorting in our dedicated guide.](/learn/advanced/geosearch.md#sorting-results-with-geopoint)

### Matching strategy

**Parameter**: `matchingStrategy`
**Expected value**: `last` or `all`
**Default value**: `last`

> Defines the strategy used to match query terms in documents. `last` returns documents containing all the query words first. If there aren't enough documents to meet the requested `limit`, it will remove one query term at a time, starting from the end of the query term.

Author: ```suggestion
Defines the strategy used to match query terms in documents. `last` returns documents containing all the query words first. If there aren't enough documents to meet the requested `limit`, it will remove one query term at a time, starting from the end of the query.
```

> Defines the strategy used to match query terms in documents. `last` returns documents containing all the query words first. If there aren't enough documents to meet the requested `limit`, it will remove one query term at a time, starting from the second last query term.

Author: ```suggestion
Defines the strategy used to match query terms in documents. `last` returns documents containing all the query words first. If there aren't enough documents to meet the requested `limit`, it will remove one query term at a time, starting from the end of the query term.
```

<CodeSamples id="search_parameter_guide_matching_strategy_1" />

With the above code sample, Meilisearch will first return documents that contain all three words. If the results don't meet the requested `limit`, it will return documents with just `big fat` followed by just `big`.

`all`, on the other hand, only returns documents that contain all the query words. Meilisearch will not match any more documents even if there aren't enough to meet the requested `limit`.

Searching for `big fat liar` will only return documents that contain all three words.

<CodeSamples id="search_parameter_guide_matching_strategy_2" />

The above code sample only returns one result compared to 476 when the `matchingStrategy` was set to `last`.

::: warning
The number of search results may vary depending on changes to the `movies` dataset.

> :::

Reviewer 3: I think we should remove mentions of specific numbers here instead of relying on the warning to dispel any confusion.

For example, we could replace this paragraph and the warning with a short text explaining that `all` will usually return fewer results than `last`.
